nixos-containers: fix atomic restart

danbst opened the pull request:
nixos containers: remove --keep-unit (revert 810680bcae1f8ca23744544e87fbf03b705e9493)
    
Fixes https://github.com/NixOS/nixpkgs/issues/43652
Fixes https://github.com/NixOS/nixpkgs/issues/16753
Alternative fix for https://github.com/NixOS/nixpkgs/pull/39717
The fix in https://github.com/NixOS/nixpkgs/pull/76719 was not enough
    
    `--keep-unit` ties systemd-nspawn container to systemd unit. When
    unit is restarted (ie atomic operation "restart", not two separate
    "stop" and "start") systemd assumes that unit didn't disappear.
    Hence machine didn't disappear. Though it may as well be systemd bug
    that machine is left in `closing` state forever.
    
    If unit is stopped (or stopped due to failure), associated machine is also stopped,
    so service can start fresh.
    
    When --keep-unit is removed, resource slice isn't changed though - it is machine.slice by default for
    nspawn containers.

cc @erikarvstedt @peterhoeg @flokli @edolstra

danbst committed:
nixos containers: remove --keep-unit (revert 810680bcae1f8ca23744544e87fbf03b705e9493)

Fixes https://github.com/NixOS/nixpkgs/issues/43652
Fixes https://github.com/NixOS/nixpkgs/issues/16753
Alternative fix for https://github.com/NixOS/nixpkgs/pull/39717
The fix in https://github.com/NixOS/nixpkgs/pull/76719 was not enough

`--keep-unit` ties systemd-nspawn container to systemd unit. When
unit is restarted (ie atomic operation "restart", not two separate
"stop" and "start") systemd assumes that unit didn't disappear.
Hence machine won't disappear.

If unit is stopped (or stopped due to failure), associated machine is also stopped,
so service can start fresh.

Resource slice isn't changed though - it is machine.slice by default for
nspawn containers.

danbst committed:
tests/imperative containers: add restart test

danbst commented:
if this will be merged, we can go on and cleanup the artifacts from old solutions - WINCH signal, mixed killmode, PID mgmt in imperative containers, restart as stop/start...

flokli commented:
cc @arianvp here.

xwvvvvwx commented:
I tested this locally and can confirm that it fixes the regression of https://github.com/NixOS/nixpkgs/issues/6212 introduced in https://github.com/NixOS/nixpkgs/commit/90a3908ec333a47232b520d8f5c7f2680f348fea.

arianvp commented:
This seems fine. though im a bit confused by why it was originally introduced.

I also don't understand why we need `machinectl poweroff` to turn off a container;  but maybe this is what you're referring to with:

> WINCH signal, mixed killmode, PID mgmt in imperative containers, restart as stop/start...


From my understanding, `systemd-nspawn` should be able to just run in the unit file, and then when the `unit` is stopped, `systemd-nspawn` should stop automatically...  E.g. make this `unit` look more like the upstream `systemd-nspawn@.service` unit

```
#  SPDX-License-Identifier: LGPL-2.1+
#
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.

[Unit]
Description=Container %i
Documentation=man:systemd-nspawn(1)
Wants=modprobe@tun.service modprobe@loop.service modprobe@dm-mod.service
PartOf=machines.target
Before=machines.target
After=network.target systemd-resolved.service modprobe@tun.service modprobe@loop.service modprobe@dm-mod.service
RequiresMountsFor=/var/lib/machines

[Service]
# Make sure the DeviceAllow= lines below can properly resolve the 'block-loop' expression (and others)
ExecStart=systemd-nspawn --quiet --keep-unit --boot --link-journal=try-guest --network-veth -U --settings=override --machine=%i
KillMode=mixed
Type=notify
RestartForceExitStatus=133
SuccessExitStatus=133
Slice=machine.slice
Delegate=yes
TasksMax=16384
@SERVICE_WATCHDOG@

# Enforce a strict device policy, similar to the one nspawn configures when it
# allocates its own scope unit. Make sure to keep these policies in sync if you
# change them!
DevicePolicy=closed
DeviceAllow=/dev/net/tun rwm
DeviceAllow=char-pts rw

# nspawn itself needs access to /dev/loop-control and /dev/loop, to implement
# the --image= option. Add these here, too.
DeviceAllow=/dev/loop-control rw
DeviceAllow=block-loop rw
DeviceAllow=block-blkext rw

# nspawn can set up LUKS encrypted loopback files, in which case it needs
# access to /dev/mapper/control and the block devices /dev/mapper/*.
DeviceAllow=/dev/mapper/control rw
DeviceAllow=block-device-mapper rw

[Install]
WantedBy=machines.target
```

arianvp commented:
So I think, by merging this change, we can completely get rid of the `machinectl stop` command, as doing `systemctl stop container@blah` should be enough to shut down the container.

flokli commented:
@danbst this looks good - can you do the above mentioned cleanups in this PR, too?

danbst commented:
@arianvp omg, there was an upstream issue for this! Interesting, they mentioned `RemainAfterExit`, I think we can't get rid of it in NixOS, as it will break NixOS restart logic.

@flokli ok, I'll try cleanup that

arianvp commented:
We'll have to keep `KillMode=mixed` by the way. it's also used upstream. I asked for the reason and this was the answer:

> michich, when the kernel wants to stop properly the system (for instance, because you pressed the power button) it sends 
SIGTERM to init, wait for some time, then powers down                                                                    
to emulate that for a container, you send SIGTERM to the main process (the init of the container)                        
then you wait a bit                                                                                                      
then you SIGKILL everything, which is the equivalent for a container of shutting down

flokli commented:
@danbst poke ;-)

danbst commented:
@flokli I'm debugging tests, they fail on `nixos-container destroy` command

danbst committed:
nixos containers: cleanups for atomic restart
* machinectl terminate during container start was a hack, probably wrong one
* use "systemctl restart" consistently for declarative and imperative containers
* fix "nixos-container terminate" command. We have to notify systemd we want
  unit to be stopped, because otherwise it will be started back due to
  restart on failure policy.
* add a bit docs to stop/terminate/destroy, to clarify which one does what
* add missing "nixos-container restart" docs

danbst commented:
Cleanups (from commit message):
```
* machinectl terminate during container start was a hack, probably wrong one
* use "systemctl restart" consistently for declarative and imperative containers
* fix "nixos-container terminate" command. We have to notify systemd we want
  unit to be stopped, because otherwise it will be started back due to
  restart on failure policy.
* add a bit docs to stop/terminate/destroy, to clarify which one does what
* add missing "nixos-container restart" docs
```
@arianvp I was not able to remove `machinectl poweroff`, for some reason it doesn't poweroff machine, but just kills it.

Also, with this PR `machinectl terminate` behavior had changed. I think I'll add release notes, to clarify this bit. `nixos-container terminate/destroy` works though.

danbst commented:
@GrahamcOfBorg test containers-imperative

flokli commented on pkgs/tools/virtualization/nixos-container/nixos-container.pl:
```diff
@@ -25,7 +25,7 @@ sub showHelp {
          [--nixos-path <path>]
          [--system-path <path>]
-         [--config <string>]
+         [--config <string>]  -- config without outer braces, for example 'services.nginx.enable = true;'
```
This is already documented in `nixos/doc/manual/administration/imperative-containers.xml`.

danbst replied:
it's fine to duplicate it here. The hands-on `--help` is too scarce.

Ma27 replied:
I agree with @danbst here. This is a common misunderstanding (which usually seems like a bug to an end-user), so mentioning this here as well is fine IMHO.

flokli commented on pkgs/tools/virtualization/nixos-container/nixos-container.pl:
```diff
@@ -34,12 +34,13 @@ sub showHelp {
+       nixos-container restart <container-name>
+       nixos-container stop <container-name>  -- shutdown container cleanly, wait until stopped
+       nixos-container terminate <container-name> -- halt container, like hard poweroff
+       nixos-container destroy <container-name> -- terminate (halt) container and remove state data
```
This should move into `nixos/doc/manual/administration/imperative-containers.xml` (if not already there). Why were things reordered?

danbst replied:
well, technically yes, but it's fine duplicating it here. No need to consult manual when stopping container.

Things were reordered to group stop/terminate/destroy commands. They do same, but with different nuances.

flokli commented on pkgs/tools/virtualization/nixos-container/nixos-container.pl:
```diff
@@ -272,6 +273,7 @@ sub isContainerRunning {
 sub terminateContainer {
     my $leader = getLeader;
+    system("systemctl", "stop", "--no-block", "container\@$containerName");
     system("machinectl", "terminate", $containerName) == 0
```
Urgh, this feels like a hack. Doesn't `machinectl terminate` by itself make the systemd unit become stopped?

danbst replied:
this is consequence of removing `--keep-unit`, I haven't found a better way to stop unit with machine.

`machinectl stop` does stop unit fine. It is because machine exits with normal result. But when it runs `terminate`, it exits with failure result, and systemd restarts unit due to `Restart=on-failure` policy.

arianvp commented:
I found a more proper fix (that cleanly shuts down the container with just
systemctl stop). We have a slight difference between the way upstream calls
systemd-nspawn. I have a patch ready and I'll test if it works tomorrow.
Wil also get rid of the WINCH hacks so that's good too.

This serves as a reminder to send that tomorrow when I have internet again
:)



On Wed, Feb 26, 2020, 23:06 Danylo Hlynskyi <notifications@github.com>
wrote:

> *@danbst* commented on this pull request.
> ------------------------------
>
> In pkgs/tools/virtualization/nixos-container/nixos-container.pl
> <https://github.com/NixOS/nixpkgs/pull/80169#discussion_r384782585>:
>
> >         nixos-container status <container-name>
>         nixos-container update <container-name>
> +       nixos-container restart <container-name>
> +       nixos-container stop <container-name>  -- shutdown container cleanly, wait until stopped
> +       nixos-container terminate <container-name> -- halt container, like hard poweroff
> +       nixos-container destroy <container-name> -- terminate (halt) container and remove state data
>
> well, technically yes, but it's fine duplicating it here. No need to
> consult manual when stopping container.
>
> Things were reordered to group stop/terminate/destroy commands. They do
> same, but with different nuances.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/80169?email_source=notifications&email_token=AAEZNI23LM57XO22RH7DDILRE3OR5A5CNFSM4KVWU6W2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCXCXNEQ#discussion_r384782585>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEZNIY7R2OBVCY2SP6GNNDRE3OR5ANCNFSM4KVWU6WQ>
> .
>

danbst commented:
@arianvp kind ping :)

arianvp commented:
My 'fix' didn't really help unfortunately.  I hit  what I think is a bug in
systemd https://github.com/systemd/systemd/issues/14961

(The fix *was setting *KillSignal=SIGRTMIN+3 with KeepUnit, such that
systemd-nspawn gracefully kills the underlying container on SIGTERM)



On Sun, Mar 1, 2020, 19:24 Danylo Hlynskyi <notifications@github.com> wrote:

> @arianvp <https://github.com/arianvp> kind ping :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/80169?email_source=notifications&email_token=AAEZNI43WFCPY22HCPMCFL3RFKR4DA5CNFSM4KVWU6W2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENNGVNY#issuecomment-593128119>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEZNIYHWP5UM3WDTRGKOUDRFKR4DANCNFSM4KVWU6WQ>
> .
>

Ma27 commented:
@danbst @arianvp may I ask if there are any updates here? :)

arianvp commented:
I'm okay with this approach for now if we are adamant on getting it out for 20.03  

If not, I would like to wait until we complete the 245 release, and see if the issues I encountered in the linked systemd issue goes away.  If we keep using `--keep-unit` with the inclusion of `--kill-signal=SIGTRM3` then `systemctl stop container@foo` should be sufficient to kill the container, without the `machinectl terminate` call

flokli commented:
I'm not sure which approach is referred to by "this", whether it's [this PR](https://github.com/NixOS/nixpkgs/pull/80169), or the one referencing it (https://github.com/NixOS/nixpkgs/pull/84608). I don't think any of these too can still be backported to 20.03, it's too late for that, and I don't want to cause breakages that late in the release cycle.

tadfisher commented:
Ping @danbst; this is still an issue in master, and it would be nice to have this fixed there at least.

maralorn commented:
Hello @arianvp and @danbst, are there any news on this one?

lIt would be so great to have some kind of fix. Today was another day where I woke up to all my containers being down after the nightly system update.
